test/hpke_test.c

```diff
@@ -0,0 +1,1410 @@
+    0x54, 0x65, 0x73, 0x74, 0x43, 0x6f, 0x6e, 0x74,
+    0x65, 0x78, 0x74
+};
+static const unsigned char fourth_export3[] = {
+    0xff, 0xaa, 0xbc, 0x85, 0xa7, 0x76, 0x13, 0x6c,
+    0xa0, 0xc3, 0x78, 0xe5, 0xd0, 0x84, 0xc9, 0x14,
+    0x0a, 0xb5, 0x52, 0xb7, 0x8f, 0x03, 0x9d, 0x2e,
+    0x87, 0x75, 0xf2, 0x6e, 0xff, 0xf4, 0xc7, 0x0e
+};
+
+static int export_only_test(void)
+{
+    /* based on RFC9180 A.7 */
+    const TEST_BASEDATA basedata = {
+        OSSL_HPKE_MODE_BASE,
+        {
+            OSSL_HPKE_KEM_ID_X25519,
+            OSSL_HPKE_KDF_ID_HKDF_SHA256,
+            OSSL_HPKE_AEAD_ID_EXPORTONLY
+        },
+        fourth_ikme, sizeof(fourth_ikme),
+        fourth_ikmepub, sizeof(fourth_ikmepub),
+        fourth_ikmr, sizeof(fourth_ikmr),
+        fourth_ikmrpub, sizeof(fourth_ikmrpub),
+        fourth_ikmrpriv, sizeof(fourth_ikmrpriv),
+        fourth_expected_shared_secret, sizeof(fourth_expected_shared_secret),
+        ksinfo, sizeof(ksinfo),
+        NULL, 0, /* no auth */
+        NULL, 0, NULL /* PSK stuff */
+    };
+    const TEST_EXPORTDATA exportdata[] = {
+        { NULL, 0, fourth_export1, sizeof(fourth_export1) },
+        { fourth_context2, sizeof(fourth_context2),
+          fourth_export2, sizeof(fourth_export2) },
+        { fourth_context3, sizeof(fourth_context3),
+          fourth_export3, sizeof(fourth_export3) },
+    };
+    return do_testhpke(&basedata, NULL, 0,
+                       exportdata, OSSL_NELEM(exportdata));
+}
+
+/*
+ * Randomly toss a coin
+ */
+static unsigned char rb = 0;
+#define COIN_IS_HEADS (RAND_bytes_ex(testctx, &rb, 1, 10) && rb % 2)
+
+/* tables of HPKE modes and suite values */
+static int hpke_mode_list[] = {
+    OSSL_HPKE_MODE_BASE,
+    OSSL_HPKE_MODE_PSK,
+    OSSL_HPKE_MODE_AUTH,
+    OSSL_HPKE_MODE_PSKAUTH
+};
+static uint16_t hpke_kem_list[] = {
+    OSSL_HPKE_KEM_ID_P256,
+    OSSL_HPKE_KEM_ID_P384,
+    OSSL_HPKE_KEM_ID_P521,
+    OSSL_HPKE_KEM_ID_X25519,
+    OSSL_HPKE_KEM_ID_X448
+};
+static uint16_t hpke_kdf_list[] = {
+    OSSL_HPKE_KDF_ID_HKDF_SHA256,
+    OSSL_HPKE_KDF_ID_HKDF_SHA384,
+    OSSL_HPKE_KDF_ID_HKDF_SHA512
+};
+static uint16_t hpke_aead_list[] = {
+    OSSL_HPKE_AEAD_ID_AES_GCM_128,
+    OSSL_HPKE_AEAD_ID_AES_GCM_256,
+    OSSL_HPKE_AEAD_ID_CHACHA_POLY1305
+};
+
+/* strings that can be used, with names or IANA codepoints */
+static char *kem_str_list[] = {
+    "P-256", "P-384", "P-521", "x25519", "x448",
+    "0x10", "0x11", "0x12", "0x20", "0x21",
+    "16", "17", "18", "32", "33"
+};
+static char *kdf_str_list[] = {
+    "hkdf-sha256", "hkdf-sha384", "hkdf-sha512",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3"
+};
+static char *aead_str_list[] = {
+    "aes-128-gcm", "aes-256-gcm", "chacha20-poly1305",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3"
+};
+/* table of bogus strings that better not work */
+static char *bogus_suite_strs[] = {
+    "3,33,3",
+    "bogus,bogus,bogus",
```

Comment:
reorder tests also dont work right?

Reply:
yep, added one of those now you mention it